onnxruntime/contrib_ops/webgpu/bert/rotary_embedding.cc

```diff
@@ -70,8 +180,22 @@ Status FusedQKRotaryEmbeddingProgram::GenerateShaderCode(ShaderHelper& shader) c
       << "  let half_rotary_dim = uniforms.cos_cache_shape[1];\n"
       << "  let bsnh = global_idx / uniforms.q_global_stride % uniforms.q_global_shape;\n"
       << "  if (bsnh[3] < half_rotary_dim) {\n"
-      << "    let pos_ids_idx = " << position_ids.BroadcastedIndicesToOffset("bsnh.xy", dummy_indices) << ";\n"
-      << "    let position_id = u32(" << position_ids.GetByOffset("pos_ids_idx") << ") + select(0u, bsnh[1], pos_ids_idx == 0u);\n"
+      << "    let batch_idx = bsnh[0];\n"
+      << "    let sequence_idx = bsnh[1];\n"
+      << "    let seqlen_i = " << seqlens.GetByOffset("batch_idx") << ";\n"
+      << "    let seqlen = u32(seqlen_i);\n"
+      << "    var position_id: u32 = 0u;\n"
+      << "    if (uniforms.first_prompt_flag == 1u) {\n"
+      << "      let total_seqlen = seqlen + 1u;\n"
+      << "      position_id = select(1u, sequence_idx, sequence_idx < total_seqlen);\n"
+      << "    } else if (uniforms.subsequent_prompt_flag == 1u) {\n"
+      << "      let total_seqlen = seqlen + 1u;\n"
+      << "      let past_seqlen = total_seqlen - uniforms.q_global_shape[1];\n"
+      << "      let cand = past_seqlen + sequence_idx;\n"
+      << "      position_id = select(1u, cand, cand < total_seqlen);\n"
+      << "    } else {\n"
+      << "      position_id = seqlen;\n"
+      << "    }\n"
```

Comment:
For line 188-198, I think you can unify it as below. It seems that past_seqlen + sequence_idx is always smaller than total_seqlen. Not sure why we need the `select` checking. In this way, we can avoid to use `first_prompt_flag `and `subsequent_prompt_flag`.
```
let total_seqlen = seqlen + 1u;
let past_seqlen = total_seqlen - uniforms.q_global_shape[1];
position_id = past_seqlen + sequence_idx;